### PR TITLE
`struct Rav1dFrameContext_task_thread:` Give fields inner mutability

### PR DESCRIPTION
**Pull request by rinon**
Adds inner mutability to the `retval` and `update_set` fields of `Rav1dFrameContext_task_thread`.

Also adds a finished flag so we don't have to inspect the `tiles` field.

### CODE REVIEW

**Review by randomPoison — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -521,9 +525,9 @@ unsafe fn drain_picture(c: &mut Rav1dContext, out: &mut Rav1dPicture) -> Rav1dRe
             c.frame_thread.next = 0 as c_int as c_uint;
         }
         drop(task_thread_lock);
-        let error = f.task_thread.retval;
+        let mut error = f.task_thread.retval.try_lock().unwrap();
         if error.is_err() {
-            f.task_thread.retval = Ok(());
+            let error = mem::replace(&mut *error, Ok(()));
```
Comment (kkysen) [inline src/lib.rs:530]: We can just unconditionally do the `mem::replace` and not need a `mut error`.

File: src/internal.rs [hunk 2]
```diff
@@ -702,8 +702,9 @@ pub(crate) struct Rav1dFrameContext_task_thread {
     pub tasks: UnsafeCell<Rav1dTasks>,
     pub init_done: AtomicI32,
     pub done: [AtomicI32; 2],
-    pub retval: Rav1dResult,
-    pub update_set: bool, // whether we need to update CDF reference
+    pub retval: Mutex<Rav1dResult>,
```
[the remark this answers is not in the document]
Comment (rinon) [reply src/internal.rs:704]: I thought about it, but it looks like that wouldn't buy us much, it's mostly accessed on error handling paths.

File: src/lib.rs [hunk 3]
```diff
@@ -489,7 +493,7 @@ unsafe fn drain_picture(c: &mut Rav1dContext, out: &mut Rav1dPicture) -> Rav1dRe
         let next: c_uint = c.frame_thread.next;
         let f: &mut Rav1dFrameData = &mut *(c.fc).offset(next as isize);
         let mut task_thread_lock = c.task_thread.delayed_fg.lock().unwrap();
-        while !f.tiles.is_empty() {
+        while !f.task_thread.finished.load(Ordering::SeqCst) {
```
[the remark this answers is not in the document]
Comment (rinon) [reply src/lib.rs:448]: I think the old code was racy here, although with no real impact. This seems to be a field we are checking and writing at the same time.

[the remark this answers is not in the document]
Comment (kkysen) [reply src/lib.rs:448]: Okay, `SeqCst` should be fine here then.  Is this something we should make a note of that `dav1d` is racy?

**Pull request merged**